accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,119 @@
+# Feature switch
+The functionality available in the .NET libraries is getting bigger and wider with every release. Applications typically don't need or want all of it, which means when an app includes the libraries as part of it, app size is unnecessarily increased. With technologies like the linker, we have the ability to remove parts of the framework which the app doesn't use. These technologies have limitations based on existing public API patterns and code behavior, sometimes they can't determine what the app really needs. This proposal adds a way to explicitly disable functionality in the framework.
+
+## Goals
+* The ability for developers to declaratively remove functionality in the framework, 3rd party libraries or final app
```

Comment:
> Libraries can fail with a sensible error message (i.e. detect when functionality they need was removed)

If the specific feature uses a public API property (as described below in the doc), then the library can easily check the value of that property and potentially fail if it requires the feature in question. In all cases it can check the runtime property value via `AppContext.TryGetSwitch`, but that is not a very nice experience. Maybe we should expose most (if not all) features via a public API property to remedy this.

> Application authors get better tooling to understand what their dependencies need

This is the hard problem. Basically the features switches are in some sense a workaround to tooling not being able to figure out the real dependencies of an application. Without them the tooling has to keep all possibilities open, but that can bring in lot of code. The feature switch is a way for the app to "help" the tooling do a better job at removing dependencies. If we could solve these automatically, then we would not need the feature switches.

We'll see what exact feature switches we'll end up with, I hope that most of them will be very easy to explain and thus easy to reason about.

Reply:
@Nirmal4G I probably don't understand what you mean by "transitive flow".

The feature switches should only be set by the application project (the one building the final app) - as that's the one which will run the linker. We don't really support running linker on anything else from the SDK right now.
Libraries and frameworks will expose feature switches by reacting to runtime properties and possibly also adding msbuild properties (to make it easier to use). I don't expect libraries (and definitely not frameworks) to actually set any feature switch - that's up to the app.

Similarly for runtime properties - these should only be set by the application (and the proposed feature switch MSBuild integration will make that automatic).

`runtimeconfig.dev` files serve different purpose (they are supposed to be the local machine configuration). I also don't think libraries/frameworks have any say to what goes into `runtimeconfig.dev`.

The above discussed "flow" of feature switches between framework/libraries/app is that if framework defines a feature switch, it should probably expose it as a public API (public static bool property) to which both libraries and app can react accordingly. Linker will see the usage of that property in all the components of the app and apply trimming accordingly as well.

Reply:
> I probably don't understand what you mean by "transitive flow"

_It was conceptual question about trimming by setting feature flags!_

Say if a lib depends on a feature and the app developer sets that switch to trim it from project and like that to several other libs or possibly in a framework pack. What will happen? The code that depends on the feature will fail at runtime.

What I was saying is to trim only when all the code paths from those dependences are safe from that feature flag! **Reflection** and compiled **RegEx** are few ones that create a problem with the linker.